### ISSUE TITLE
Cool.rand isn't very rand

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1685]
[The docs](https://docs.perl6.org/routine/rand) (and my own expectations) say that `$n.rand` returns a random number between `0` and `$n`. In reality, it looks like instead a random number between `0` and `1` is generated and then scaled to `$n`'s size.

As a result, all of the generated numbers are clustered near `$n`. The `%pow` in the code below collects the magnitude of all the generated random numbers:

```
<Zoffix_> m: my %pow; for ^100_000 { %pow{1e300.rand.comb(/\d+/).tail}++ }; say %pow
<camelia> rakudo-moar 5981f81b6: OUTPUT: «{294 => 1, 295 => 10, 296 => 86, 297 => 856, 298 => 8999, 299 => 90048}␤»

<Zoffix_> m: my %pow; for ^100_000 { %pow{(10**300).rand.comb(/\d+/).tail}++ }; say %pow
<camelia> rakudo-moar 5981f81b6: OUTPUT: «{294 => 2, 295 => 4, 296 => 95, 297 => 929, 298 => 8941, 299 => 90029}␤»

<Zoffix_> m: my %pow; for ^10_000 { %pow{1e30.rand.comb(/\d+/).tail}++ }; say %pow
<camelia> rakudo-moar 5981f81b6: OUTPUT: «{26 => 8, 27 => 90, 28 => 893, 29 => 9009}␤»
```

- Perl 5 does the same thing (so if we keep the same behaviour, then docs need to be changed)
- Ruby returns Ints and the docs claim the generated Ints are between 0 and max

**Comment by colomon** [comment]
I'm not going to tell you that the algorithm is actually correct -- I haven't even looked at it -- but aren't these results basically what a perfect random number generator would yield?

That is to say: if you do a (correctly random) (10**2).rand, the result should have two digits (ie it's from 10-99) about nine times as often as it has one digit (0-10).  For (10**3).rand, the results have three digits (100-999) ~90% of the time, two digits (10-99) ~9% of the time, and one digit (0-9) ~0.9% of the time.

The numbers you're getting around following that pattern, no?

**Comment by zoffixznet** [comment]
Ah, right! Thanks 😄

**Comment by colomon** [comment]
I should perhaps have indicated that it's still possible the algorithm *is* wrong.  Back when we first implemented is-prime (on Niecza), I remember that it was important for the RNG to generate full randomness for that algorithm.  If all you do is generate a random Num between 0 and 1 and then multiply it by `10**300`, only first 15 digits or so of the number will be properly random -- the rest will not be very random at all.  That's fine if your result is a Num, but bad news if you want a truly random large Int.  (I remember conjuring up a giant random number by generating a series of random bytes and pasting them together to make a big enough Int.)

**Comment by colomon** [comment]
> my %pow; for ^100_000 { %pow{(10**300).rand % 256}++ }; say %pow;
    {0 => 100000}

Indeed, this looks very wrong...

**Comment by colomon** [comment]
Well... `(10**300).rand` is Num.  If that's correct, then this result is correct.

**Comment by colomon** [comment]
For what it's worth, `(^(10**300)).roll` does generate numbers which look (to above simple test) as if the last byte is indeed random.

**Comment by samcv** [comment]
Yes this is very non-random. You can tell if you put it in base-16:
```
(^(10**300)).rand.base(16)
CDDE783439D2F00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

**Comment by zoffixznet** [comment]
> You can tell if you put it in base-16:

Would that be just the available precision *in that particular generated Num*?

```
<Zoffix_> m: say 10e300.base: 16
<camelia> rakudo-moar c8364d696: OUTPUT: «EEEA5D5004981800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000␤»
```

In other words, yes, you don't get the full range of numbers do to the limitations of floating point format, but you still can get a double that's say `1e3`. It's just it's extremely unlikely.

I think colomon's original explanation is solid. There's a million *times* more numbers between 1e300 and 1e294 than between 1e294 and 0, so it makes sense that the former would occur much more frequently if picked pseudo-randomly.

**Comment by colomon** [comment]
Just to be clear, I have no idea if the algorithm we use is capable of generating, as you say, `1e3`.  I don't know what the underlying rand routine looks like.

On the other hand, if it were actually evenly weighted across the integers from `0` to `1e300` and you had been generating a trillion such random numbers a second since the Big Bang, odds are still fantastically good you wouldn't have gotten a number smaller than `1e100`.  (If you think of this generation as rolling 299 ten-sided dice, you'd have to roll at least 199 zeros in a row to get a number that small!)  It's simply not something you can check out with a practical test.

**Comment by ab5tract** [comment]
>  (If you think of this generation as rolling 299 ten-sided dice, you'd have to roll at least 199 zeros in a row to get a number that small!) 

I'm not a probability expert, but my understanding of `(10**300).rand` is that it is meant to mimic a _single_ roll of a `10**300`-sided die. In that case, every value between 0 and 10**300 should have an equal chance of occurring. I don't know how to square this with the idea that values below a given threshold would have low chance of occuring for a `now - BigBang` running process.

But, as I said, I'm not a probability expert.

From my experience with recent work with `Range.rand`, though, there seems to be a point at which the value returned by the underlying `rand` operation becomes non-viable when used as an input into Raku's number tower.

```
> r '(1e-15).FatRat ==> say()'
0
```

Conversion to our highest precision data type fails at a certain (floating) point. Is it such a leap to imagine that this implies a skew when combined with large enough endpoints?

I don't know. I'm not an expert.

**Comment by niner** [comment]
@ab5tract the low chance comes from the simple fact that there are so many more numbers between 1e100 and 1e300 than there are between 0 and 1e100. So while every individual number has an equal chance of appearing, you just won't see one lower than 1e100. Same as you won't see one between 1e300-1e100 and 1e300. The same is true for any **specific** number range of size 1e100. There are 1e200 of them! So the chance for any individual one is just 1/1e200. That's an unimaginably small number. Is it possible that a number < 1e100 gets picked? Absolutely! Will it happen? Absolutely not! (And yes, my own intuition is fighting my rational mind about this, but it's a lost cause).

Randomness and large numbers are both things humans deal badly with. If you combine them the badness multiplies.

If floating point values are involved in your random number generation at any point, then they limit the actual randomness and the number range you can get. If it's a float between 0 and 1 you'll get about 62 bits of actual randomness and this 62 bit number (~19 decimal digits) gets scaled to the range you requested. That's a fundamental limit of this process.

If you really want full randomness, you need to generate enough random bytes to fill the whole number. For a range from 0 to 1e300 you'd need 997 bits, so 125 random bytes. But then, you are calling a method on a Num, so you should expect the result to also be a Num and those will only have 64 bits to store the result, so you will lose some (or a lot of) precision.

We do however have the "good" way implemented as well. On ranges `pick` and `roll` will both do the right thing using the nqp::rand_I op to get the random number which actually does fill the bigint with these random bytes:

```
> raku -e 'say (1e100).rand.Int.base(16)'
E5C6A908D6206000000000000000000000000000000000000000000000000000000000000000000000

> raku -e 'say (^10**100).roll.base(16)'
B8E348AD8983F2D5AEFA762403246037FCD873BC33EF5991F186F881746AE465E5F9E0FED84129CE3DD
```

In 6.e you won't need a Range and can do this just with Ints:
```
> raku -e 'use v6.e.PREVIEW; say (10**100).roll.base(16)'
EDC46D514FCAF803A10595C86951754CD840461705BE86B28F038D59930EE79F127537D84393695B036
```

With this I'd say this issue can be closed. We simply cannot store more random bits in a Num than it has space for and there are good alternatives that you can use.

**Comment by ab5tract** [comment]
Well put! The difference in "volume" as we move up in magnitudes makes perfect sense.

It was also mentioned earlier in this thread and almost clicked for me, but I then promptly forgot it by the time I was writing the response. Hopefully it will stay in place a bit longer this time :)
